# Day 5 In-class Assignment: Savings for Everyone
![Snowballs](https://www.palmettocitizens.org/images/personal-savings3.jpg)

### <p style="text-align: right;"> &#9989; Kaitlyn Watson
<p style="text-align: right;"> &#9989; Owen Babington, Lorenzo Durancharris, Povilas Kvederas, David Callaway

## Learning Goals:

By the end of this assignment you should be able to:

* Review code that uses `if`, `elif`, and `else` statments and understand the logical flow of the program
* Write `if`/`elif`/`else` statements in Python.
* Create and use basic functions in Python.


---
## Part 1: Understanding `if` statements in Python

As you saw in the pre-class assignment, `if` statements provide an mechanism for "**branching**" to occur in your code. Depending on whether or not something is determined to be true, your code can pursue one branch, or another, or do nothing at all. The `else` statement allows your code to execute *something* even if none of the `if` or `elif` statements you write are found to be true. By combining these sorts of commands in a variety of ways we can create a very complex system of branches in our code.

&#9989;&nbsp; **As a group, review and run** the following bit of example code.

&#9989;&nbsp; **As a group, add comments** to explain what various parts of the code are doing. 

In [1]:
some_numbers = [2, -5, 16, 17, -6, 23, 0]  #Created a list of some numers

for i in range(len(some_numbers)):  #Create a for loop to loop through all the numbers in the list above. 
    number = float(some_numbers[i]) #Establish a variable that uses the float funtion to convert the list above into floating numbers.
    if number < 0:                         #Create an if else loop to establish whether or not the numbers in the list are real
        print('Square root is not real.')
    else:
        print('Square root of',number,'is',number**(1/2))
        
for number in some_numbers:     #Create a second for loop
    if number > 0:              #Create an if, elif, else statement to determine whether the number is positive, negative, or zero.
        print(number,'is positive.')
    elif number < 0:
        print(number,'is negative.')
    else:
        print('The number is zero.')

Square root of 2.0 is 1.4142135623730951
Square root is not real.
Square root of 16.0 is 4.0
Square root of 17.0 is 4.123105625617661
Square root is not real.
Square root of 23.0 is 4.795831523312719
Square root of 0.0 is 0.0
2 is positive.
-5 is negative.
16 is positive.
17 is positive.
-6 is negative.
23 is positive.
The number is zero.


&#9989;&nbsp; **Question**: What does `float(some_numbers[i])` do?

<font size="+3">&#9998;</font> This part of the code converts the list of some_numbers into floating point numbers.

&#9989;&nbsp; **Question**: What does the `len()` function do?

<font size="+3">&#9998;</font> The len() function directs the code to run through the number values in the list above. 

&#9989;&nbsp; **Question**: What does the `range()` function do?

<font size="+3">&#9998;</font> The range() function tells the for loop to run the code through a specific interval of values. In this case, the length of the list some_numbers.

---
## Part 2: Writing your own `if` statements

Sometimes in computational modeling and data science, you need to be able parse a set of numbers or data and execute particular tasks based on the how those numbers compare to each other. Conditional statements that use things like `<`, `>`, `>=`, or `==` combined with logical operators like `and` and `or` provide an excellent mechanism for comparing values! Try the following:

&#9989;&nbsp; **Write a program** that compares three variables v1, v2, and v3 (initial values are provide below) and does the following using `if` statements:

1. If v1 is equal to v2, print "woohoo!"
2. If v1 is the smallest value of the three (i.e., v1 is less than both v2 and v3), print out "v1 is tiny!"
3. If v1 is the smallest value of the three, add a nested if statement to print out a statement saying whether v2 or v3 is the largest value.

Try several values of v1, v2 and v3 and ensure that you see the correct behavior. 

Make sure you work together with your group to solve this. **Plan it out on the whiteboard!**

**Note**: Although you learned briefly about functions in your pre-class assignment, don't worry about using functions just yet!

In [14]:
# Put your code here. Initial values for v1, v2, and v3 are provided.
# Make sure to test out new values to ensure your code works as intended!

v1 = 1
v2 = 2
v3 = 3

if v1==v2:
    print("woohoo!")

if v1 < v2 and v1 < v3:
    print('v1 is tiny!')

if v1 < v2 and v1 < v3:
    if v2 < v3:
        print("The largest number is v2")
    else:
        print("The largest number is v3")



v1 is tiny!
The largest number is v2


&#9989;&nbsp; **Question**: What happens in your code if either v2 or v3 is less than v1? 

<font size="+3">&#9998;</font> If v2 or v3 is than v1 there is no output in the loop. Nothing prints. 

---
## Part 3: Continued Investment

You're back at your bank job, calculating the total amount of money people save in their retirement accounts. Previously, your boss asked you to calculate the different amount of money people would save if there was one interest rate for people that started saving early (under 30) and another interest rate (twice the original interest rate) if they started saving later (30 or older). 

Your boss used your calculations to justify to *their* boss that offering a "better interest rate for older investors" was a good marketing strategy. Your boss's boss accepted the plan with one modification. Rather than give people over 30 double the interest rates, the interest rates would be structured in the following way:
- People that start to save when they are between **18-24 years old** will receive an interest rate of **6.5%**
- People that start to save when they are between **25-31 years old** will receive an interest rate of **8.5%**
- People that start to save when they are between **32-39 years old** will receive an interest rate of **11.0%**
- People that start to save when they are **40 years or older** will receive an interest rate of **17.0%**

(Also, it is assumed that **everyone will retire at 65**)

The bank rolled out the plan, and it worked *really well.* Now, there are dozens of new applications for savings accounts, and your boss says you need to calculate the expected savings for all of these new applications. 

You will need to **write a piece of code that will go through each of these new applications and calculate each individual's expected retirement savings.  Your code should make a list of all of the expected retirement savings.**

**Note:** In finding a solution to this problem, we **strongly recommend** that you use your code from the previous assignment involving compound interest.

### &#9989;&nbsp; 3.1 Plan out your code

You and your groups members are **expected to use your whiteboard**. 

*Things to think about:*
- How will you go through each application?
- What variables will change for each application?
- What calculations do you need to make for each application?
- How can you use the code you wrote in the previous compound interest assignment?

1. Create a loop in order to go through each client and their application.
2. The savings per year, age of the client, and the output for final retirement savings.
3. Multiply the savings by year with the interest rate and calculate the final retirement savings balance.
4. We can use the code in the previous assignment within the loop that we create for each client. 

### &#9989;&nbsp; 3.2 Write out your code

Below are lists of all of the new applications, including the clients name, the amount they will save each year, and the age at which they're starting to save. 

**Write your code below. Store the final retirement savings in the list `final_retirement_savings`**

In [16]:
# New applications
client_name = ["Michael Mitchell","Ashley Moore","Fred Walker","Walter Wood","Jesse Gonzalez","Phillip Sanders","Marie Jenkins","Annie Brown","Christopher King","Carl Phillips","Betty Perez","Ashley Hernandez","Carlos Davis","Richard Jackson","Jean Cox","Sara Allen","Peter Butler","Justin Collins","Kevin Miller","Andrew Russell"]
savings_per_year = [4680,4680,4320,5880,5280,5640,4680,5520,5040,4200,4920,5400,5280,4560,5040,4440,5640,5040,4800,4680]
start_savings_age = [20,31,35,39,40,37,26,21,36,25,25,36,33,27,35,31,28,19,31,34]

final_retirement_savings = []

In [55]:
# Put your code here

for i in range(len(client_name)):
    age= start_savings_age[i]
    if age>= 18 and age<= 24:
        Int= 1.065
    elif age>= 25 and age<= 31:
        Int= 1.085
    elif age>= 32 and age<= 39:
        Int= 1.11
    else:
        Int= 1.17
        
    balance = 0
    while age<= 65:
        balance += balance*Int
        balance += savings_per_year[i]
        age += 1
    final_retirement_savings.append(balance)
    print(client_name[i], 'will have', final_retirement_savings[i])
    
    
        
    

Michael Mitchell will have 1232411.0046493162
Ashley Moore will have 901843.8408388686
Fred Walker will have 586539.2599626718
Walter Wood will have 556804.7885654417
Jesse Gonzalez will have 1809786.9576993773
Phillip Sanders will have 640489.4471283826
Marie Jenkins will have 1383794.2695972258
Annie Brown will have 1359711.7181622384
Christopher King will have 626042.2150136253
Carl Phillips will have 1351625.3176398627
Betty Perez will have 1583332.514949554
Ashley Hernandez will have 670759.5160860276
Carlos Davis will have 854939.4092672352
Richard Jackson will have 1238481.4426254185
Jean Cox will have 684295.8032897835
Sara Allen will have 855595.438744568
Peter Butler will have 1406604.6035264293
Justin Collins will have 1418520.621486254
Kevin Miller will have 924968.0418860192
Andrew Russell will have 694108.021814457


In [54]:
#proof that the above code is correct for all clients 

for i in range (len(client_name)):
    number = float(savings_per_year[i])
    age = start_savings_age[i]
    if age>=18 and age<=24:
        new_fund=number+number*1.065
        number=new_fund
        final_retirement_savings.append(new_fund)
        
print(client_name[i],final_retirement_savings[i])


#Based on Day 3 in class code 

Andrew Russell 694108.021814457


In [27]:
# Print out the final savings for all of the new applications

for kk in range(len(final_retirement_savings)):
    print("{} will save ${:.2f} million by the time they retire".format(client_name[kk],final_retirement_savings[kk]/1.0e6))

Michael Mitchell will save $1.23 million by the time they retire
Ashley Moore will save $0.90 million by the time they retire
Fred Walker will save $0.59 million by the time they retire
Walter Wood will save $0.56 million by the time they retire
Jesse Gonzalez will save $1.81 million by the time they retire
Phillip Sanders will save $0.64 million by the time they retire
Marie Jenkins will save $1.38 million by the time they retire
Annie Brown will save $1.36 million by the time they retire
Christopher King will save $0.63 million by the time they retire
Carl Phillips will save $1.35 million by the time they retire
Betty Perez will save $1.58 million by the time they retire
Ashley Hernandez will save $0.67 million by the time they retire
Carlos Davis will save $0.85 million by the time they retire
Richard Jackson will save $1.24 million by the time they retire
Jean Cox will save $0.68 million by the time they retire
Sara Allen will save $0.86 million by the time they retire
Peter Butler

IndexError: list index out of range

&#9989;&nbsp; **Question**: Which client will retire with the *largest* amount of money in their savings account? When did they start saving? How much money did they add every year?

<font size="+3">&#9998;</font> Jesse Gonzalez will have |1.81 million saved by the time they retire. They began saving when they 40 and thet added $5280 each year. 

&#9989;&nbsp; **Question**: Which client will retire with the *smallest* amount of money in their savings account? When did they start saving? How much money did they add every year?

<font size="+3">&#9998;</font> Christopher King has the lowest amount of money at 0.63 million. He began saving at 36 and saved $5040 each year. 

&#9989;&nbsp; **Question**: Compare the people that retired with the most and least amount of money. What was the leading cause of the differences between the two? Is there anything that the person with the least amount of money saved could have done differently?

<font size="+3">&#9998;</font> The leading difference between the largest and the smallest is the age. Jesse had a larger interest rate or 17 percent becuase he started saving at a later age, while Christopher started saving at 36 and only had an interest rate of 11 percent. 

---
### &#128721; STOP
Check in with all of your group members to make sure you all understand everything to this point!

---
## Part 4: Writing functions

Functions fulfill a similar role as loops: they allow us to avoid repeating the same code over and over again. The rule of thumb we will use for this class is: 

### If you need to repeat some piece of code three or more times turn it into a function

In today's assignment and a previous assignment, you used a block of code that calculated the total amount of money saved for retirement. Let's use this bit of code to get some practice writing functions.

### &#9989;&nbsp; 4.1: Writing a function
**Write a function that calculates the total amount of money saved in a retirement account.** You should be able to insert this function directly into the code you wrote in the previous exercise.

In [60]:
# Put your code for your new function here
def retirement_account (age, money):
    if age>= 18 and age<= 24:
        Int= 1.065
    elif age>= 25 and age<= 31:
        Int= 1.085
    elif age>= 32 and age<= 39:
        Int= 1.11
    else:
        Int= 1.17
    
    balance = 0
    while age<= 65:
        balance += balance*Int
        balance += money
        age += 1
    return(balance)

### &#9989;&nbsp; 4.2: Using your Function

**In the cell below, put your code from Part 3.2. Use the function you wrote to calculate the total amount of money saved in the retirement account.**

In [61]:
# Put your code here


    

TypeError: retirement_account() missing 1 required positional argument: 'Int'

In [62]:
# Print out the final savings for all of the new applications

for kk in range(len(final_retirement_savings)):
    print("{} will save ${:.2f} million by the time they retire".format(client_name[kk],final_retirement_savings[kk]/1.0e6))

Michael Mitchell will save $1.23 million by the time they retire
Ashley Moore will save $0.90 million by the time they retire
Fred Walker will save $0.59 million by the time they retire
Walter Wood will save $0.56 million by the time they retire
Jesse Gonzalez will save $1.81 million by the time they retire
Phillip Sanders will save $0.64 million by the time they retire
Marie Jenkins will save $1.38 million by the time they retire
Annie Brown will save $1.36 million by the time they retire
Christopher King will save $0.63 million by the time they retire
Carl Phillips will save $1.35 million by the time they retire
Betty Perez will save $1.58 million by the time they retire
Ashley Hernandez will save $0.67 million by the time they retire
Carlos Davis will save $0.85 million by the time they retire
Richard Jackson will save $1.24 million by the time they retire
Jean Cox will save $0.68 million by the time they retire
Sara Allen will save $0.86 million by the time they retire
Peter Butler

IndexError: list index out of range

---

## Congratulations, you're done!

Submit this assignment by uploading it to the course Desire2Learn web page.  Go to the "In-class assignments" folder, find the appropriate submission link, and upload it there.

See you next class!

&#169; Copyright 2021,  Michigan State University Board of Trustees